vm/systemSmartContracts/liquidStaking.go

```diff
@@ -0,0 +1,244 @@
+	Marshalizer            marshal.Marshalizer
+	Hasher                 hashing.Hasher
+	EpochNotifier          vm.EpochNotifier
+}
+
+// NewLiquidStakingSystemSC creates a new liquid staking system SC
+func NewLiquidStakingSystemSC(args ArgsNewLiquidStaking) (*liquidStaking, error) {
+	if check.IfNil(args.Eei) {
+		return nil, vm.ErrNilSystemEnvironmentInterface
+	}
+	if len(args.DelegationMgrSCAddress) < 1 {
+		return nil, fmt.Errorf("%w for delegation manager sc address", vm.ErrInvalidAddress)
+	}
+	if len(args.EndOfEpochAddress) < 1 {
+		return nil, fmt.Errorf("%w for end of epoch address", vm.ErrInvalidAddress)
+	}
+	if len(args.LiquidStakingSCAddress) < 1 {
+		return nil, fmt.Errorf("%w for liquid staking sc address", vm.ErrInvalidAddress)
+	}
+	if check.IfNil(args.Marshalizer) {
+		return nil, vm.ErrNilMarshalizer
+	}
+	if check.IfNil(args.Hasher) {
+		return nil, vm.ErrNilHasher
+	}
+	if check.IfNil(args.EpochNotifier) {
+		return nil, vm.ErrNilEpochNotifier
+	}
+
+	l := &liquidStaking{
+		eei:                      args.Eei,
+		delegationMgrSCAddress:   args.DelegationMgrSCAddress,
+		endOfEpochAddr:           args.EndOfEpochAddress,
+		liquidStakingSCAddress:   args.LiquidStakingSCAddress,
+		gasCost:                  args.GasCost,
+		marshalizer:              args.Marshalizer,
+		hasher:                   args.Hasher,
+		liquidStakingEnableEpoch: args.EpochConfig.EnableEpochs.BuiltInFunctionOnMetaEnableEpoch,
+	}
+	log.Debug("liquid staking: enable epoch", "epoch", l.liquidStakingEnableEpoch)
+
+	args.EpochNotifier.RegisterNotifyHandler(l)
+
+	return l, nil
+}
+
+// Execute calls one of the functions from the delegation contract and runs the code according to the input
+func (l *liquidStaking) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	l.mutExecution.RLock()
+	defer l.mutExecution.RUnlock()
+
+	err := CheckIfNil(args)
+	if err != nil {
+		l.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	if !l.flagLiquidStaking.IsSet() {
+		l.eei.AddReturnMessage("liquid staking contract is not enabled")
+		return vmcommon.UserError
+	}
+
+	switch args.Function {
+	case core.SCDeployInitFunctionName:
+		return l.init(args)
+	case "claimDelegatedPosition":
+		return l.claimDelegatedPosition(args)
+	case "claimRewardsFromPosition":
+		return l.claimRewardsFromDelegatedPosition(args)
+	case "reDelegateRewardsFromPosition":
+		return l.reDelegateRewardsFromPosition(args)
+	case "unDelegateWithPosition":
+		return l.unDelegateWithPosition(args)
+	case "returnPosition":
+		return l.returnPosition(args)
+	}
+
+	l.eei.AddReturnMessage(args.Function + " is an unknown function")
+	return vmcommon.UserError
+}
+
+func (l *liquidStaking) init(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !bytes.Equal(args.CallerAddr, l.liquidStakingSCAddress) {
+		l.eei.AddReturnMessage("invalid caller")
+		return vmcommon.UserError
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		l.eei.AddReturnMessage("not a payable function")
```

Comment:
"function is not payable in eGLD" to keep the same message format used in all the other methods from this file